lib/elixir/test/elixir/code_normalizer_test.exs

```diff
@@ -0,0 +1,476 @@
+      assert quoted_to_string(quote(do: -1)) == "-1"
+      assert quoted_to_string(quote(do: +(+1))) == "+(+1)"
+      assert quoted_to_string(quote(do: !(foo > bar))) == "!(foo > bar)"
+      assert quoted_to_string(quote(do: @foo(bar))) == "@foo bar"
+      assert quoted_to_string(quote(do: identity(&1))) == "identity(&1)"
+    end
+
+    test "access" do
+      assert quoted_to_string(quote(do: a[b])) == "a[b]"
+      assert quoted_to_string(quote(do: a[1 + 2])) == "a[1 + 2]"
+      assert quoted_to_string(quote(do: (a || [a: 1])[:a])) == "(a || [a: 1])[:a]"
+      assert quoted_to_string(quote(do: Map.put(%{}, :a, 1)[:a])) == "Map.put(%{}, :a, 1)[:a]"
+    end
+
+    test "keyword list" do
+      assert quoted_to_string(quote(do: [a: a, b: b])) == "[a: a, b: b]"
+      assert quoted_to_string(quote(do: [a: 1, b: 1 + 2])) == "[a: 1, b: 1 + 2]"
+      assert quoted_to_string(quote(do: ["a.b": 1, c: 1 + 2])) == "[\"a.b\": 1, c: 1 + 2]"
+    end
+
+    test "interpolation" do
+      assert quoted_to_string(quote(do: "foo#{bar}baz")) == ~S["foo#{bar}baz"]
+    end
+
+    test "bit syntax" do
+      ast = quote(do: <<1::8*4>>)
+      assert quoted_to_string(ast) == "<<1::8*4>>"
+
+      ast = quote(do: @type(foo :: <<_::8, _::_*4>>))
+      assert quoted_to_string(ast) == "@type foo :: <<_::8, _::_*4>>"
+
+      ast = quote(do: <<69 - 4::bits-size(8 - 4)-unit(1), 65>>)
+      assert quoted_to_string(ast) == "<<69 - 4::bits-size(8 - 4)-unit(1), 65>>"
+
+      ast = quote(do: <<(<<65>>), 65>>)
+      assert quoted_to_string(ast) == "<<(<<65>>), 65>>"
+
+      ast = quote(do: <<65, (<<65>>)>>)
+      assert quoted_to_string(ast) == "<<65, (<<65>>)>>"
+
+      ast = quote(do: for(<<(a::4 <- <<1, 2>>)>>, do: a))
+      assert quoted_to_string(ast) == "for <<(a::4 <- <<1, 2>>)>> do\n  a\nend"
+    end
+
+    test "charlist" do
+      assert quoted_to_string(quote(do: [])) == "[]"
+      assert quoted_to_string(quote(do: 'abc')) == "'abc'"
+    end
+
+    test "string" do
+      assert quoted_to_string(quote(do: "")) == ~S/""/
+      assert quoted_to_string(quote(do: "abc")) == ~S/"abc"/
+      assert quoted_to_string(quote(do: "#{"abc"}")) == ~S/"#{"abc"}"/
+    end
+
+    test "last arg keyword list" do
+      assert quoted_to_string(quote(do: foo([]))) == "foo([])"
+      assert quoted_to_string(quote(do: foo(x: y))) == "foo(x: y)"
+      assert quoted_to_string(quote(do: foo(x: 1 + 2))) == "foo(x: 1 + 2)"
+      assert quoted_to_string(quote(do: foo(x: y, p: q))) == "foo(x: y, p: q)"
+      assert quoted_to_string(quote(do: foo(a, x: y, p: q))) == "foo(a, x: y, p: q)"
+
+      assert quoted_to_string(quote(do: {[]})) == "{[]}"
+      assert quoted_to_string(quote(do: {[a: b]})) == "{[a: b]}"
+      assert quoted_to_string(quote(do: {x, a: b})) == "{x, [a: b]}"
+      assert quoted_to_string(quote(do: foo(else: a))) == "foo(else: a)"
+      assert quoted_to_string(quote(do: foo(catch: a))) == "foo(catch: a)"
+    end
+  end
```

Comment:
Also, let's add some tests to make sure that if we call string_to_quoted_with_comments with the token_metadata+literal_encoder and then we call quoted_to_algebra, the output will be as close as possible to the one from the formatter. In particular, here are some tests of things that should be kept as is:

* code comments
* escape on strings, such as `"\x0A"` should remain as is and not `"\n"`
* floats and integers, such as 1.0e1 and 0b101

We can probably have these tests in their own describe block.